test/e2e_node/device_plugin_test.go

```diff
@@ -934,7 +934,10 @@ func testDevicePluginNodeReboot(f *framework.Framework, pluginSockDir string) {
 		// simulate node reboot scenario by removing pods using CRI before kubelet is started. In addition to that,
 		// intentionally a scenario is created where after node reboot, application pods requesting devices appear before the device plugin pod
 		// exposing those devices as resource has restarted. The expected behavior is that the application pod fails at admission time.
-		framework.It("Keeps device plugin assignments across node reboots (no pod restart, no device plugin re-registration)", framework.WithFlaky(), func(ctx context.Context) {
+		// NOTE: this is flaky because all pods race to go running when kubelet restarts. So application pod will run with device plugion pod.
+		// It's possible, even though unlikely, that the device plugin pod manages to go running *and* the kubelet reports device healthy
+		// when the application pod enters admission. Hence the flakes.
```

Comment:
Is it possible to postpone admission of the application pod until device plugin pod is registered? It would make the whole setup much more predictable.

Reply:
we will need to add the admission retry loop in kubelet, which we talk about since quite some time (e.g. https://github.com/kubernetes/kubernetes/issues/119248 - note this is not limited to device plugins even though device plugins are a prominent use case) and add some form of dependency tracking between pods; or perhaps, something like scheduling gates (https://kubernetes.io/blog/2022/12/26/pod-scheduling-readiness-alpha/) but for kubelet admission: "don't even try admission until this condition is met; retry later" + a timeout.